src/road_gui.cpp

```diff
@@ -1093,144 +1093,137 @@ static void ShowRoadDepotPicker(Window *parent)
+				if (roadstoptype == ROADSTOP_BUS && spec->stop_type != ROADSTOPTYPE_PASSENGER && spec->stop_type != ROADSTOPTYPE_ALL) continue;
+				return true;
+			}
+		}
+		return false;
+	}
+
+	static bool IsClassChoice(const RoadStopClass &cls)
+	{
+		return !IsWaypointClass(cls) && GetIfClassHasNewStopsByType(&cls, roadstoptype, _cur_roadtype);
+	}
+
+	bool HasClassChoice() const override
+	{
+		return std::count_if(std::begin(RoadStopClass::Classes()), std::end(RoadStopClass::Classes()), IsClassChoice);
+	}
 
-	typedef GUIList<RoadStopClassID, std::nullptr_t, StringFilter &> GUIRoadStopClassList; ///< Type definition for the list to hold available road stop classes.
+	int GetSelectedClass() const override { return _roadstop_gui.sel_class; }
+	void SetSelectedClass(int id) const override { _roadstop_gui.sel_class = this->GetClassIndex(id); }
 
-	static const uint EDITBOX_MAX_SIZE = 16; ///< The maximum number of characters for the filter edit box.
+	StringID GetClassName(int id) const override
+	{
+		const auto *rsc = this->GetClass(id);
+		if (!IsClassChoice(*rsc)) return STR_NULL;
+		return rsc->name;
+	}
 
-	static Listing   last_sorting;           ///< Default sorting of #GUIRoadStopClassList.
-	static Filtering last_filtering;         ///< Default filtering of #GUIRoadStopClassList.
-	static const std::initializer_list<GUIRoadStopClassList::SortFunction * const> sorter_funcs;   ///< Sort functions of the #GUIRoadStopClassList.
-	static const std::initializer_list<GUIRoadStopClassList::FilterFunction * const> filter_funcs; ///< Filter functions of the #GUIRoadStopClassList.
-	GUIRoadStopClassList roadstop_classes;   ///< Available road stop classes.
-	StringFilter string_filter;              ///< Filter for available road stop classes.
-	QueryString filter_editbox;              ///< Filter editbox.
+	int GetSelectedType() const override { return _roadstop_gui.sel_type; }
+	void SetSelectedType(int id) const override { _roadstop_gui.sel_type = id; }
+
+	StringID GetTypeName(int cls_id, int id) const override
+	{
+		const auto *spec = this->GetSpec(cls_id, id);
+		if (!IsRoadStopEverAvailable(spec, roadstoptype == ROADSTOP_BUS ? STATION_BUS : STATION_TRUCK)) return STR_NULL;
+		return (spec == nullptr) ? STR_STATION_CLASS_DFLT_ROADSTOP : spec->name;
+	}
 
-	void EnsureSelectedClassIsVisible()
+	bool IsTypeAvailable(int cls_id, int id) const override
 	{
-		uint pos = 0;
-		for (auto rs_class : this->roadstop_classes) {
-			if (rs_class == _roadstop_gui_settings.roadstop_class) break;
-			pos++;
+		const auto *spec = this->GetSpec(cls_id, id);
+		return IsRoadStopAvailable(spec, roadstoptype == ROADSTOP_BUS ? STATION_BUS : STATION_TRUCK);
+	}
+
+	void DrawType(int x, int y, int cls_id, int id) const override
+	{
+		const auto *spec = this->GetSpec(cls_id, id);
+		if (spec == nullptr) {
+			StationPickerDrawSprite(x, y, roadstoptype == ROADSTOP_BUS ? STATION_BUS : STATION_TRUCK, INVALID_RAILTYPE, _cur_roadtype, _roadstop_gui.orientation);
```

Comment:
Looks like you copied and pasted this, but should this be `INVALID_ROADTYPE`?

```suggestion
			StationPickerDrawSprite(x, y, roadstoptype == ROADSTOP_BUS ? STATION_BUS : STATION_TRUCK, INVALID_ROADTYPE, _cur_roadtype, _roadstop_gui.orientation);
```

Reply:
Nope, this is rabbit hole where StationPickerDrawSprite is used for both rail and road stations.